### ISSUE TITLE
implement pagination in explorer frontend

### ISSUE THREAD

**Issue by zaibon** [threefoldtech/zos#722]

**Comment by DylanVerstraete** [comment]
Not quite sure on what we try to achieve.

For now, fetching all the data seems reasonable since we don't have a big userbase. In the future if our usebase gets biggers we can implement caching of the node requests. The node informating does not get updated every minute I think so we can have a cache time of like 3-5 minutes or something.

If we don't implement caching following needs to happen:

1. pagenation requests in the nodes table
  - this creates another issue on the filtering of node ids and resources (since the filter will only be applied on 1 page of nodes)
  - if we reactivly implement filtering on the nodes, then we still have to scan our entire collection for subsets, this will cause alot of overhead if this is done by multiple users at the same time.

2. compute the node stats in the backend and cache this request (since this request fetches all the nodes to compute the stats)

**Comment by zaibon** [comment]
First thing to solve is to make sure the explorer can load all the nodes from the explorer API. Which is not the case today: https://github.com/threefoldtech/zos/blob/77b9939bffb2de0fe09657f96194b9842d3e379c/tools/explorer/frontend/src/services/tfService.js#L39-L46

Now doing this brings other questions indeed. How to deal with a huge amount of nodes ?
At the moment all the computation of statistics etc is done on the client side. Which I think is a good thing, we do not want to put that burden on the server.
So following that train of though it seems we should always load all the nodes in the frontend store.

When we would reach a too big amount of nodes that loading them all in memory is not an option anymore. then I think we need to rethink a bit how the server works. Caching is a solution but is is always quite hard to get right. Specially that nodes can get updated data at any time. Not only every X minutes.


So long story short, for now I think we can just make all the nodes load in the background when the frontend loads. And then monitor how things evolve with the growing number of node ?

**Comment by DylanVerstraete** [comment]
Personally, I have never seen an application loading pieces of data for a small period of time until everything is loaded. This will make the statistics out of sync at the start and then the frontend will be doing alot more re-renders then is needed.

Since the request for the nodes on production is only 31.0 KB it doesn't seem like it outweighs the effort here. If we would have like over 10k nodes even then it will still be doable. 

Maybe we can research if we render the page server side we can reduce overhead?

**Comment by LeeSmet** [comment]
- Page rendering server side is a bad idea imo, don't think we get anything from that. On the contrary. If we do a query, we'll need to send it to the backend, the backend renders it, and sends it back. So we do a request for every query, thus increase the load on the server by a lot for no real benefit.

- Queries on node id/farm id could be handled by the backend. There are already filters for this in the db. This also means that if you share a query, you don't waste the bandwith when it is loaded again on sending useless nodes.

All in all, if the request size is that small, I think its doable to load everything. If we want response design, indeed we could load an initial amount of 100 or so nodes, and then in the background load additional nodes. This gives at least a couple of seconds to load the entire thing before a user would actually do a query. And if a user immediatly queries while still streaming, we can always show a loading spinner in the node table imo.

**Comment by zaibon** [comment]
>All in all, if the request size is that small, I think its doable to load everything. If we want response design, indeed we could load an initial amount of 100 or so nodes, and then in the background load additional nodes. This gives at least a couple of seconds to load the entire thing before a user would actually do a query. And if a user immediatly queries while still streaming, we can always show a loading spinner in the node table imo.

This is what I meant by loading all the nodes. Just load a first reasonable batch once then pull the rest in the background. That allow to have a responsive UI and still load everything.
Going further in this idea. It would be possible then refresh each nodes over time based on their last update time or something.

> Maybe we can research if we render the page server side we can reduce overhead?

I do not like this. We want to make the load of the server light, no on the client.
